### PR TITLE
Add arm64 simulator slice to binary builds.

### PR DESCRIPTION
**Pull request by ryanwilson**
This re-architects things to split by platform and build for
architectures together instead of having to lipo them together. Still
need to fix Carthage and test a full release.

### CODE REVIEW

File: ReleaseTooling/Sources/ZipBuilder/CocoaPodUtils.swift [hunk 1]
```diff
@@ -543,10 +543,15 @@ enum CocoaPodUtils {
       """)
     case let .success(output):
       let version = output.components(separatedBy: ".")
+      guard version.count >= 2 else {
+        fatalError("Failed to parse CocoaPods version: \(version)")
```
Comment (ryanwilson) [inline ReleaseTooling/Sources/ZipBuilder/CocoaPodUtils.swift:547]: I believe this is what is causing the occasional failure on CI - I ran into it locally once by chance.

Comment (paulb777) [reply ReleaseTooling/Sources/ZipBuilder/CocoaPodUtils.swift:547]: Hmm, I wonder how that could ever happen?

File: ReleaseTooling/Sources/ZipBuilder/FrameworkBuilder.swift [hunk 2]
```diff
@@ -303,12 +333,8 @@ struct FrameworkBuilder {
       } catch {
         fatalError("Error while enumerating files \(frameworkPath): \(error.localizedDescription)")
       }
-      var libPath = frameworkPath.appendingPathComponent(actualFramework)
```
Comment (paulb777) [inline ReleaseTooling/Sources/ZipBuilder/FrameworkBuilder.swift:306]: Needs testing that dynamic frameworks still get built right?

File: ReleaseTooling/Sources/ZipBuilder/FrameworkBuilder.swift [hunk 3]
```diff
@@ -519,55 +552,56 @@ struct FrameworkBuilder {
...
         archive from Terminal with:
-        rm \(thinArchive.path)
+        rm \(slicedFramework.path)
         """)
       }
     }
     return (xcframework, carthageFramework)
   }
 
+  /// Assembles a `.framework` file from a fat archive and other necessary information.
```
Comment (paulb777) [inline ReleaseTooling/Sources/ZipBuilder/FrameworkBuilder.swift:593]: s/file/folder/ ?

File: ReleaseTooling/Sources/ZipBuilder/FrameworkBuilder.swift [hunk 4]
```diff
@@ -634,94 +655,96 @@ struct FrameworkBuilder {
...
+                                                          includingPropertiesForKeys: nil)
+            .compactMap { $0.path }
+          let destSwiftModuleDir = destModuleDir
+            .appendingPathComponent(swiftModule.lastPathComponent)
+          for file in files {
+            let fileURL = URL(fileURLWithPath: file)
+            let projectDir = swiftModule.appendingPathComponent("Project")
+            if fileURL.lastPathComponent == "Project",
+              fileManager.directoryExists(at: projectDir) {
+              // The Project directory (introduced with Xcode 11.4) already exist, only copy in
```
Comment (paulb777) [inline ReleaseTooling/Sources/ZipBuilder/FrameworkBuilder.swift:701]: s/exist/exists/

File: ReleaseTooling/Sources/ZipBuilder/FrameworkBuilder.swift [hunk 5]
```diff
@@ -634,94 +655,96 @@ struct FrameworkBuilder {
     }
   }
 
-  private func makeSwiftModuleMap(thinArchives: [Architecture: URL], destination: URL) -> Bool {
+  // TODO: Verify if the array is needed, or we can pass in an individual framework for searching.
+  // Why do we need to parse each one? The argument passed in previously was previously the thin
+  // binary slices, not a single URL.
+  private func makeSwiftModuleMap(slicedFramework: URL, destination: URL) -> Bool {
```
Comment (paulb777) [inline ReleaseTooling/Sources/ZipBuilder/FrameworkBuilder.swift:661]: This is something that really could use a unit test ... but not right now

File: ReleaseTooling/Sources/ZipBuilder/FrameworkBuilder.swift [hunk 6]
```diff
@@ -847,15 +866,16 @@ struct FrameworkBuilder {
...
 
     // Package a normal .framework with the given slices.
     let frameworkDir = platformFrameworksDir.appendingPathComponent(fromFolder.lastPathComponent)
-    packageFramework(withName: framework,
-                     fromFolder: fromFolder,
-                     thinArchives: slices,
-                     destination: frameworkDir,
-                     moduleMapContents: moduleMapContents)
+    fatalError("TODO: Fix this. Likely take in an XCFramework and unwrap it to be a framework")
+//    packageFramework(withName: framework,
```
Comment (paulb777) [inline ReleaseTooling/Sources/ZipBuilder/FrameworkBuilder.swift:874]: TODO

**Review by paulb777 — COMMENTED** [review]
Nice improvements!

File: ReleaseTooling/Sources/ZipBuilder/FrameworkBuilder.swift (hunk 5 above)
Comment (ryanwilson) [reply ReleaseTooling/Sources/ZipBuilder/FrameworkBuilder.swift:661]: Agreed 😄

**Comment by ryanwilson** [comment]
Fixing CI, please wait for review.

**Comment by ryanwilson** [comment]
Fixed.

**Review by paulb777 — APPROVED** [review]
Thanks!

**Comment by paulb777** [comment]
Ideally, we should understand the Database and Firestore cron failures before making a release with this

**Comment by ryanwilson** [comment]
> Ideally, we should understand the Database and Firestore cron failures before making a release with this

SGTM, I'll get working on those now and reach out to Gran.

**Comment by paulb777** [comment]
Should be good to go after a clean CI run of #6986

**Comment by ryanwilson** [comment]
I'm going to land this to unblock building a new RC to include arm64 slices for the simulator, and will update it if I see any issues the the generated artifact.

**Pull request merged**